### DISCUSSION TITLE
Raintree Research Questions

### DISCUSSION THREAD

**Discussion by jeremybeal11** [pokt-network/pocket#339] [Q&A] [answered]
Going to use this discussion post as a repository of questions I have when reading up on RainTree. Sorry in advance if some of these are noob-ish: 

1.  In the RainTree Spec [doc](https://github.com/pokt-network/pocket-network-protocol/tree/main/p2p#2-algorithm-gossip), it mentioned a `replacement` node, but it never referenced how it comes into the picture when nodes are gossiping and adjusting its peer list.  I'm sure the answer is simple, but can it be clarified here how it occurs when nodes are talking to each other? 
2.  When the originator node(on L4) is sending out the ACK requests to the left and right branches(on L3), is that request going out to the nodes(on L3) at the same time, or will there be a delay(I.e .001 ms)?
3.  Followup Q(for Q2) - assuming the originator(L4) receives both ACK from the two branches(on L3) and demotes itself to the next layer, how does the originator choose which of the four total nodes(connected to the two parents on L3) to choose for it's left and right branches? or it just selects the nodes that are closest in distance?  

@Olshansk, I hope some of these questions make sense hahaha

**Comment by Olshansk** [comment, marked as answer]
Similar to #345, also wanted to point to the original [RainTree presentation](https://github.com/pokt-network/pocket/files/9853354/raintree.pdf) put together by @BenVanGithub who can help answer questions as well.

--- 

> 1. In the RainTree Spec [doc](https://github.com/pokt-network/pocket-network-protocol/tree/main/p2p#2-algorithm-gossip), it mentioned a replacement node, but it never referenced how it comes into the picture when nodes are gossiping and adjusting its peer list. I'm sure the answer is simple, but can it be clarified here how it occurs when nodes are talking to each other?

If timeout happens => need a replacement node.

What is replacement node => next node in the "ordered list of actors" + 1 level lower


> 2. When the originator node(on L4) is sending out the ACK requests to the left and right branches(on L3), is that request going out to the nodes(on L3) at the same time, or will there be a delay(I.e .001 ms)?

I wouldn't consider `0.001ms` a delay in this context. In the theoretical sense, it's the same time. In terms of practicality & implementation, I think that's hard to achieve.

> 3. Followup Q(for Q2) - assuming the originator(L4) receives both ACK from the two branches(on L3) and demotes itself to the next layer, how does the originator choose which of the four total nodes(connected to the two parents on L3) to choose for it's left and right branches? or it just selects the nodes that are closest in distance?

I believe the implementation we have is the best source for this at the moment: https://github.com/pokt-network/pocket/blob/main/p2p/raintree/peers_manager.go

<img width="1204" alt="Screen Shot 2022-11-14 at 10 14 33 AM" src="https://user-images.githubusercontent.com/1892194/201735424-192f8be6-b54a-4f7c-9865-ec8be472a913.png">

Based on either the protocol level actors (i.e. staked nodes) or the full peer list (full nodes and protocol level actors), we sort everyone based on their advertised P2P public address. This is the list/order used to achieve structured gossip. 

**[Unsure]** There's still an open question as to whether it'll be just the actors or the full peer list, but the fact that pocket has thousands of nodes staked allows us to only use protocol level actors for structured gossip, and there can be a long-term incentive around this as well.
